### PR TITLE
Makes PlatformInformationProvider aware of the browser default route …

### LINKED ISSUE #87897: Navigator 2.0 doesn't update url to specified initial route
## Steps to Reproduce

1. Run `flutter create bug`.
2. Update the files as follows: 
    main.dart -> https://gist.github.com/hrishikesh-kadam/78f784338c3e85a5d66fb3b4537c81d2
3. `flutter run -d chrome`

**Expected results:** 
The location bar should update to `localhost:port/#/book/1`

**Actual results:** 
The location bar stays on `localhost:port`, not even `localhost:port/` or `localhost:port/#/`

What is the right way to set the initial route in the `Router` widget?
The example used here is referred from the article [Learning Flutter’s new navigation and routing system](https://medium.com/flutter/learning-flutters-new-navigation-and-routing-system-7c9068155ade).

<details>
  <summary>Logs</summary>

`$ flutter run -d chrome --verbose`
```
[  +48 ms] executing: uname -m
[  +23 ms] Exit code 0 from: uname -m
[        ] x86_64
[   +4 ms] executing: [/home/hrk/snap/flutter/common/flutter/] git -c
log.showSignature=false log -n 1 --pretty=format:%H
[   +6 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] f4abaa0735eba4dfd8f33f73363911d63931fe03
[        ] executing: [/home/hrk/snap/flutter/common/flutter/] git tag --points-at
f4abaa0735eba4dfd8f33f73363911d63931fe03
[   +9 ms] Exit code 0 from: git tag --points-at f4abaa0735eba4dfd8f33f73363911d63931fe03
[        ] 2.2.3
[   +3 ms] executing: [/home/hrk/snap/flutter/common/flutter/] git rev-parse --abbrev-ref
--symbolic @{u}
[   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/stable
[        ] executing: [/home/hrk/snap/flutter/common/flutter/] git ls-remote --get-url
origin
[   +3 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +58 ms] executing: [/home/hrk/snap/flutter/common/flutter/] git rev-parse --abbrev-ref
HEAD
[   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] stable
[  +39 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping
update.
[  +41 ms] executing: /mnt/common-data/softwares/android/sdk/linux/platform-tools/adb
devices -l
[   +6 ms] executing:
/home/hrk/snap/flutter/common/flutter/bin/cache/artifacts/fuchsia/tools/x64/ffx target list
--format s
[   +8 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping
update.
[  +44 ms] executing: /mnt/common-data/softwares/android/sdk/linux/platform-tools/adb
devices -l
[        ] executing:
/home/hrk/snap/flutter/common/flutter/bin/cache/artifacts/fuchsia/tools/x64/ffx target list
--format s
[   +8 ms] Skipping pub get: version match.
[  +84 ms] Generating
/mnt/common-data/study-material/flutter/bugs/initial_route/bug/android/app/src/main/java/io
/flutter/plugins/GeneratedPluginRegistrant.java
[  +79 ms] List of devices attached
[   +1 ms] List of devices attached
[  +41 ms] Launching lib/main.dart on Chrome in debug mode...
[  +76 ms] Updating assets
[  +55 ms] Waiting for connection from debug service on Chrome...
[  +64 ms] No devices found.
[        ] No devices found.
[  +12 ms] <- reset
[   +3 ms] /home/hrk/snap/flutter/common/flutter/bin/cache/dart-sdk/bin/dart
--disable-dart-dev
/home/hrk/snap/flutter/common/flutter/bin/cache/artifacts/engine/linux-x64/frontend_server.
dart.snapshot --sdk-root /home/hrk/snap/flutter/common/flutter/bin/cache/flutter_web_sdk/
--incremental --target=dartdevc --debugger-module-names --experimental-emit-debug-metadata
-DFLUTTER_WEB_AUTO_DETECT=true --output-dill
/tmp/flutter_tools.ZSDQGC/flutter_tool.TFAJIX/app.dill --libraries-spec
file:///home/hrk/snap/flutter/common/flutter/bin/cache/flutter_web_sdk/libraries.json
--packages
/mnt/common-data/study-material/flutter/bugs/initial_route/bug/.dart_tool/package_config.js
on -Ddart.vm.profile=false -Ddart.vm.product=false --enable-asserts --track-widget-creation
--filesystem-root /tmp/flutter_tools.ZSDQGC/flutter_tools.JPTYYX --filesystem-scheme
org-dartlang-app --initialize-from-dill
build/b1b715402d823b7fd5c2b68d2edcb2ce.cache.dill.track.dill --platform
file:///home/hrk/snap/flutter/common/flutter/bin/cache/flutter_web_sdk/kernel/flutter_ddc_s
dk_sound.dill --sound-null-safety
[   +5 ms] <- compile org-dartlang-app:/web_entrypoint.dart
[+13131 ms] Waiting for connection from debug service on Chrome... (completed in 13.2s)
[        ] Synced 28.3MB.
[        ] <- accept
[        ] Caching compiled dill
[ +159 ms] Using Google Chrome 92.0.4515.131 

[ +461 ms] [CHROME]:
[        ] [CHROME]:DevTools listening on
ws://127.0.0.1:42703/devtools/browser/29cbf343-0667-478a-9cb4-d5d3ecf4581f
[ +157 ms] DwdsInjector: Received request for entrypoint at
http://localhost:40379/main_module.bootstrap.js
[   +3 ms] MetadataProvider: Loading debug metadata...
[  +10 ms] MetadataProvider: Loaded debug metadata (sound null safety)
[   +5 ms] DwdsInjector: Injected debugging metadata for entrypoint at
http://localhost:40379/main_module.bootstrap.js
[+1254 ms] ChromeProxyService: Initializing expression compiler for
main_module.bootstrap.js with sound null safety: true
[  +50 ms] DevHandler: Debug service listening on ws://127.0.0.1:45237/2njNtXyFeUU=/ws

[  +16 ms] This app is linked to the debug service: ws://127.0.0.1:45237/2njNtXyFeUU=/ws
[   +5 ms] Debug service listening on ws://127.0.0.1:45237/2njNtXyFeUU=/ws
[        ] 💪 Running with sound null safety 💪
[   +2 ms] 🔥  To hot restart changes while running, press "r" or "R".
[        ] For a more detailed help message, press "h". To quit, press "q".
```

`$ flutter analyze`
```
Analyzing bug...                                                        
No issues found! (ran in 1.0s)
```

`$ flutter doctor -v`
```
[✓] Flutter (Channel stable, 2.2.3, on Linux, locale en_IN)
    • Flutter version 2.2.3 at /home/hrk/snap/flutter/common/flutter
    • Framework revision f4abaa0735 (5 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /mnt/common-data/softwares/android/sdk/linux
    • Platform android-31, build-tools 31.0.0
    • ANDROID_SDK_ROOT = /mnt/common-data/softwares/android/sdk/linux
    • Java binary at: /snap/android-studio/current/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2
    • pkg-config version 0.29.1

[✓] Android Studio (version 2020.3)
    • Android Studio at /snap/android-studio/current/android-studio
    • Flutter plugin version 59.0.2
    • Dart plugin version 203.8292
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] Android Studio (version 4.2)
    • Android Studio at /snap/android-studio/113/android-studio
    • Flutter plugin version 58.0.1
    • Dart plugin version 202.8531
    • Java version OpenJDK Runtime Environment (build 11.0.8+0-b944-P17168821)

[✓] IntelliJ IDEA Community Edition (version 2021.1)
    • IntelliJ at /snap/intellij-idea-community/310
    • Flutter plugin installed
    • Dart plugin version 211.7727

[✓] IntelliJ IDEA Community Edition (version 2021.2)
    • IntelliJ at /snap/intellij-idea-community/315
    • Flutter plugin installed
    • Dart plugin version 212.4746.57

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Linux
    • Chrome (web)    • chrome • web-javascript • Google Chrome 92.0.4515.131

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by chunhtai**
…and move the reporting logic from Router to RouteInformation provider

Previously Router checks PlatformRouteInformationProvider.value to decide whether the generated RouteInformation needed to be reported back to PlatformRouteInformationProvider. This assumes the PlatformRouteInformationProvider.value is in sync with the browser.

However, this may not always be the case. The browser initial route name is always '/' and developer can provide a different initial RouteInformation to the PlatformRouteInformationProvider constructor. In fact, Router has no way to know whether it should report or not. The pr makes it that Router will always report to RouteInformationProvider with a flag `RouteInformationReportingType`. It will be RouteInformationProvider to decide what to do with the RouteInformation.

fixes https://github.com/flutter/flutter/issues/87897

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/router.dart [hunk 1]
```diff
@@ -471,22 +468,26 @@ class Router<T> extends StatefulWidget {
...
-  // Don't report regardless of route information changes.
-  ignore,
+/// The [Router]'s intention when it reports a new [RouteInformation] to the
+/// [RouteInformationProvider].
+///
+/// See also:
+///
+///  * [RouteInformationProvider.routerReportsNewRouteInformation]: which is
+///    called by the router when it has a new route information to report.
+enum RouteInformationReportingType {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:478]: nit: the names in this enum are inconsistent: implicit is an adjective, neglecting as a verb form, and navigation is a noun. We should probably be more consistent here.

File: packages/flutter/lib/src/widgets/router.dart [hunk 2]
```diff
@@ -471,22 +468,26 @@ class Router<T> extends StatefulWidget {
...
+  /// The router does not have a specific intention.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:479]: This probably deserves a little more information. Why or when does the router not have a specific intention?

File: packages/flutter/lib/src/widgets/router.dart [hunk 3]
```diff
@@ -471,22 +468,26 @@ class Router<T> extends StatefulWidget {
...
+  implicit,
+  /// The [RouteInformation] is generated during the [Router.neglect].
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:481]:
```suggestion
  /// The accompanying [RouteInformation] were generated during a [Router.neglect] call.
```

File: packages/flutter/lib/src/widgets/router.dart [hunk 4]
```diff
@@ -471,22 +468,26 @@ class Router<T> extends StatefulWidget {
...
+  neglecting,
+  /// The [RouteInformation] is generated during the [Router.navigate].
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:483]: similar to above.

File: packages/flutter/lib/src/widgets/router.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/router.dart:478]: Maybe just neglect and navigate for the last two. Not exactly sure what we want to indicate with "implicit", so not sure what to use there.

File: packages/flutter/lib/src/widgets/router.dart [hunk 5]
```diff
@@ -471,22 +468,26 @@ class Router<T> extends StatefulWidget {
...
+/// The [Router]'s intention when it reports a new [RouteInformation] to the
+/// [RouteInformationProvider].
+///
+/// See also:
+///
+///  * [RouteInformationProvider.routerReportsNewRouteInformation]: which is
+///    called by the router when it has a new route information to report.
+enum RouteInformationReportingType {
+  /// The router does not have a specific intention.
+  implicit,
```
Comment (mdebbar) [inline packages/flutter/lib/src/widgets/router.dart:480]: Should we also explain the effect of this type? Like does it report route information changes or not?

**Comment by chunhtai** [comment]
@goderbauer  @mdebbar  updated comment, PTAL

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/router.dart [hunk 6]
```diff
@@ -1325,19 +1312,21 @@ abstract class RouteInformationProvider extends ValueListenable<RouteInformation
...
+  /// state of [RouteInformationProvider] subclass with the `routeInformation`.
+  /// For example, [PlatformRouteInformationProvider] uses this property to
+  /// decide whether to push or replace the browser history entry with the new
+  /// `routeInformation`.
   ///
   /// For more information on how [Router] determines a navigation event, see
   /// the "URL updates for web applications" section in the [Router]
   /// documentation.
-  void routerReportsNewRouteInformation(RouteInformation routeInformation, {bool isNavigation = true}) {}
+  void routerReportsNewRouteInformation(RouteInformation routeInformation, {required RouteInformationReportingType type}) {}
```
Comment (nguyenxndaidev) [inline packages/flutter/lib/src/widgets/router.dart:1329]: @chunhtai This change breaks the backward compatibility. Could we have a default value here? I guess, the default value should be `RouteInformationReportingType.none`, right?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/router.dart:1329]: yes, that is a reasonable default.

Comment (nguyenxndaidev) [reply packages/flutter/lib/src/widgets/router.dart:1329]: Thanks for your very prompt confirmation!
Do you plan to update the signature with a default value to next Flutter release version?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/router.dart:1329]: Can you open an issue for this? it can help us triage and prioritize it

Comment (nguyenxndaidev) [reply packages/flutter/lib/src/widgets/router.dart:1329]: Here you are https://github.com/flutter/flutter/issues/96136

**Pull request merged**